File: apps/nvim/lua/ctok/format.lua
--- Format ctok JSON results into human-readable lines for display.
local M = {}

--- Format a token number as "500", "1.5k", "1.5M".
---@param n number
---@return string
local function fmt_tokens(n)
  if n >= 1e6 then
    return string.format("%.1fM", n / 1e6)
  elseif n >= 1000 then
    return string.format("%.1fk", n / 1000)
  end
  return tostring(n)
end

--- Format a USD amount.
---@param v number
---@return string
local function fmt_usd(v)
  if v < 0.01 then
    return string.format("$%.4f", v)
  end
  return string.format("$%.2f", v)
end

--- Effort level → label with indicator.
---@param effort string
---@return string
local function effort_label(effort)
  local map = { low = "🟢 Low", medium = "🟡 Medium", high = "🟠 High", xhigh = "🔴 X-High" }
  return map[effort] or ("⚪ " .. (effort or "?"))
end

--- Format check result into display lines.
---@param result table Parsed JSON from ctok check --json
---@return string[]
function M.check_lines(result)
  local lines = {}
  local function push(line) table.insert(lines, line or "") end

  push("╭─ Token Estimate ─────────────────────────────────╮")
  push("")

  local est = result.estimate or {}
  local inp = est.input or {}
  local out = est.output or {}
  push(string.format("  Input:    %s  (range %s – %s)",
    fmt_tokens(inp.expected or 0), fmt_tokens(inp.min or 0), fmt_tokens(inp.max or 0)))
  push(string.format("  Output:   %s  (range %s – %s)",
    fmt_tokens(out.expected or 0), fmt_tokens(out.min or 0), fmt_tokens(out.max or 0)))
  push(string.format("  Confidence: %s", est.confidence or "?"))
  push("")

  local cost = result.cost or {}
  push(string.format("  Cost:     %s  (range %s – %s)",
    fmt_usd(cost.totalUsd or 0),
    fmt_usd((cost.totalUsdRange or {}).min or 0),
    fmt_usd((cost.totalUsdRange or {}).max or 0)))
  push(string.format("  Input:    %s   Output: %s",
    fmt_usd(cost.inputUsd or 0), fmt_usd(cost.outputUsd or 0)))
  push("")

  local rec = result.recommendation or {}
  local effort = (rec.effort or {}).effort or "?"
  local model = (rec.model or {}).model or "?"
  push(string.format("  Effort:   %s", effort_label(effort)))
  push(string.format("  Model:    %s", model))
  if (rec.effort or {}).reason then
    push(string.format("  Why:      %s", rec.effort.reason))
  end
  if (rec.model or {}).reason then
    push(string.format("            %s", rec.model.reason))
  end
  push("")

  local suggestions = result.suggestions or {}
  if #suggestions > 0 then
    push("  Suggestions:")
    for _, s in ipairs(suggestions) do
      push(string.format("  • [%s] %s", s.severity or "info", s.title or ""))
      if s.detail and s.detail ~= "" then
        push(string.format("    %s", s.detail))
      end
    end
    push("")
  end

  push("╰──────────────────────────────────────────────────╯")
  return lines
end

--- Format refine result into display lines.
---@param result table Parsed JSON from ctok refine --json
---@return string[]
function M.refine_lines(result)
  local lines = {}
  local function push(line) table.insert(lines, line or "") end

  push("╭─ Refined Prompt ─────────────────────────────────╮")
  push("")
  push(string.format("  Saved: %d tokens  (%.0f%%)",
    result.savedTokens or 0, result.savedPct or 0))
  push("")
  push("  ─ Refined ─")
  push("")
  for part in (result.refined or ""):gmatch("[^\n]+") do
    push("  " .. part)
  end
  push("")
  push("  ─ Passes Applied ─")
  local passes = result.passes or {}
  if #passes > 0 then
    for _, p in ipairs(passes) do
      if (p.saved or 0) > 0 then
        push(string.format("  • %-26s −%d tokens", p.name or "?", p.saved))
      end
    end
  else
    push("  (none)")
  end
  if result.specificityScore then
    push("")
    push(string.format("  Specificity: %d/100  →  %d/100",
      result.specificityBefore or 0, result.specificityScore))
  end
  push("")
  push("╰──────────────────────────────────────────────────╯")
  return lines
end

--- Format scan result into display lines.
---@param result table Parsed JSON from ctok scan --json
---@return string[]
function M.scan_lines(result)
  local lines = {}
  local function push(line) table.insert(lines, line or "") end

  push("╭─ Project Scan ───────────────────────────────────╮")
  push("")
  push(string.format("  Root:    %s", result.root or "."))
  push(string.format("  Type:    %s", result.projectType or "unknown"))
  push(string.format("  Files:   %d  (%s estimated tokens)",
    result.totalFiles or 0, fmt_tokens(result.estimatedTokens or 0)))
  push("")

  local by_ext = result.byExtension or {}
  local exts = {}
  for ext, data in pairs(by_ext) do
    table.insert(exts, { ext = ext, tokens = data.tokens or 0, files = data.files or 0 })
  end
  table.sort(exts, function(a, b) return a.tokens > b.tokens end)

  if #exts > 0 then
    push("  By extension:")
    for i, e in ipairs(exts) do
      if i > 10 then break end
      push(string.format("  %-12s %4d files   %s tokens",
        e.ext, e.files, fmt_tokens(e.tokens)))
    end
    push("")
  end

  local heavy = result.topHeavyFiles or {}
  if #heavy > 0 then
    push("  Heaviest files:")
    for i, f in ipairs(heavy) do
      if i > 8 then break end
      local short = f.path or ""
      if #short > 50 then short = "…" .. short:sub(-49) end
      push(string.format("  %-52s %s", short, fmt_tokens(f.tokens or 0)))
    end
    push("")
  end

  local excl = result.excluded or {}
  if (excl.files or 0) > 0 then
    push(string.format("  Excluded: %d files  (%s tokens saved)",
      excl.files, fmt_tokens(excl.tokens or 0)))
    push("")
  end

  push("╰──────────────────────────────────────────────────╯")
  return lines
end

return M

File: apps/slack/src/format.ts
import type { KnownBlock } from "@slack/bolt";
import type { CheckResult, ScanResult, RefineOutput } from "./types";

export function fmtTokens(n: number): string {
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`;
  if (n >= 1_000) return `${(n / 1_000).toFixed(1)}k`;
  return String(n);
}

export function fmtUsd(n: number): string {
  if (n < 0.01) return `$${n.toFixed(4)}`;
  return `$${n.toFixed(2)}`;
}

export function effortEmoji(effort: string): string {
  switch (effort) {
    case "low": return "🟢";
    case "medium": return "🟡";
    case "high": return "🟠";
    case "xhigh": return "🔴";
    default: return "⚪";
  }
}

export function buildCheckBlocks(r: CheckResult): KnownBlock[] {
  const { estimate, recommendation, suggestions, cost } = r;
  const effort = recommendation.effort.effort;
  const model = recommendation.model.model;

  const blocks: KnownBlock[] = [
    {
      type: "header",
      text: { type: "plain_text", text: "⚡ ctok - Token Estimate", emoji: true },
    },
    {
      type: "section",
      fields: [
        { type: "mrkdwn", text: `*Input tokens*\n${fmtTokens(estimate.input.min)}-${fmtTokens(estimate.input.max)} (est. *${fmtTokens(estimate.input.expected)}*)` },
        { type: "mrkdwn", text: `*Output tokens*\n${fmtTokens(estimate.output.min)}-${fmtTokens(estimate.output.max)}` },
        { type: "mrkdwn", text: `*Total cost*\n${fmtUsd(cost.totalUsd)} (${fmtUsd(cost.totalUsdRange.min)}-${fmtUsd(cost.totalUsdRange.max)})` },
        { type: "mrkdwn", text: `*Confidence*\n${estimate.confidence}` },
      ],
    },
    { type: "divider" },
    {
      type: "section",
      text: {
        type: "mrkdwn",
        text: `${effortEmoji(effort)} *${effort}* effort → \`${model}\`\n_${recommendation.model.reason}_`,
      },
    },
  ];

  if (suggestions.length > 0) {
    const tips = suggestions
      .slice(0, 3)
      .map((s) => `• *${s.title}*: ${s.detail}`)
      .join("\n");
    blocks.push({ type: "divider" });
    blocks.push({
      type: "section",
      text: { type: "mrkdwn", text: `*Suggestions*\n${tips}` },
    });
  }

  blocks.push({
    type: "context",
    elements: [{ type: "mrkdwn", text: "_Powered by <https://ctok-cli.github.io/ctok|ctok> - all analysis runs locally_" }],
  });

  return blocks;
}

export function buildRefineBlocks(r: RefineOutput, original: string): KnownBlock[] {
  const saved = fmtTokens(r.tokensSaved);
  return [
    {
      type: "header",
      text: { type: "plain_text", text: "✂️ ctok - Refined Prompt", emoji: true },
    },
    {
      type: "section",
      text: { type: "mrkdwn", text: `*Tokens saved:* ~${saved} (${r.savedPct.toFixed(0)}% reduction)` },
    },
    { type: "divider" },
    {
      type: "section",
      text: { type: "mrkdwn", text: `*Refined prompt:*\n\`\`\`${r.refined.slice(0, 2900)}\`\`\`` },
    },
    {
      type: "context",
      elements: [{ type: "mrkdwn", text: "_Powered by <https://ctok-cli.github.io/ctok|ctok>_" }],
    },
  ];
}

export function buildScanBlocks(r: ScanResult, directory: string): KnownBlock[] {
  const topFiles = r.topHeavyFiles
    .slice(0, 8)
    .map((f) => `\`${f.path}\` - ${fmtTokens(f.tokens)} tok`)
    .join("\n");

  return [
    {
      type: "header",
      text: { type: "plain_text", text: "🔍 ctok - Project Scan", emoji: true },
    },
    {
      type: "section",
      fields: [
        { type: "mrkdwn", text: `*Directory*\n\`${directory}\`` },
        { type: "mrkdwn", text: `*Project type*\n${r.projectType}` },
        { type: "mrkdwn", text: `*Total files*\n${r.totalFiles}` },
        { type: "mrkdwn", text: `*Total tokens*\n${fmtTokens(r.estimatedTokens)}` },
      ],
    },
    { type: "divider" },
    {
      type: "section",
      text: { type: "mrkdwn", text: `*Heaviest files:*\n${topFiles || "_No files found_"}` },
    },
    {
      type: "context",
      elements: [{ type: "mrkdwn", text: "_Powered by <https://ctok-cli.github.io/ctok|ctok>_" }],
    },
  ];
}

export function buildErrorBlock(message: string): KnownBlock[] {
  return [
    {
      type: "section",
      text: { type: "mrkdwn", text: `❌ *ctok error:* ${message}` },
    },
  ];
}

export function buildHelpBlocks(): KnownBlock[] {
  return [
    {
      type: "header",
      text: { type: "plain_text", text: "⚡ ctok - Claude Token Estimator", emoji: true },
    },
    {
      type: "section",
      text: {
        type: "mrkdwn",
        text: [
          "*Commands:*",
          "`/ctok check <prompt>` - Estimate tokens, cost, and get a model recommendation",
          "`/ctok refine <prompt>` - Run the 7-pass refiner to trim tokens and improve clarity",
          "`/ctok scan <path>` - Scan a directory and report its token footprint",
          "`/ctok help` - Show this message",
          "",
          "*Options (append to any command):*",
          "`--model haiku-4-5|sonnet-4-6|opus-4-7` - Override model for cost calculation",
          "`--plan free|pro|max5x|max20x` - Override plan for quota estimates",
          "`--task bug-fix|feature|refactor|review|...` - Hint for model recommendation",
        ].join("\n"),
      },
    },
    {
      type: "context",
      elements: [{ type: "mrkdwn", text: "<https://ctok-cli.github.io/ctok|ctok-cli.github.io/ctok> · <https://github.com/ctok-cli/ctok|GitHub>" }],
    },
  ];
}
